File: HW/HW8/HW8.py
class HW8:
    def longestPalindrome(self, s):
        char_count = {}
        
        ## Count the frequency of each character in the string
        for char in s:
            char_count[char] = char_count.get(char, 0) + 1

        longest_length = 0
        odd = False
        
        for count in char_count.values():
            
            ## If the character count is even, we can form a palindrome
            if count % 2 == 0:
                longest_length += count
                
            ## If it's odd, add its count - 1 to the longest length because an odd character
            ## cannot form a palindrome
            else:
                longest_length += count - 1
                odd = True
        
        ## If there's at least one odd character, we can place that character in the middle of the palindrome
        ## So we add +1 to the length
        if odd:
            longest_length += 1
        
        return longest_length

hw8 = HW8()
test1 = "abccccdd"
test2 = "speediskey"

print(hw8.longestPalindrome(test1))
print(hw8.longestPalindrome(test2))

File: HW/HW8/HW8.java
import java.util.HashMap;

public class HW8 {
    public int longestPalindrome(String s) {
        int longest_length = 0;
        HashMap<Character, Integer> charCount = new HashMap<>();

        // Count the frequency of each character
        for (char c : s.toCharArray()) {
            charCount.put(c, charCount.getOrDefault(c, 0) + 1);
        }
        
        // Checks for odd characters
        boolean odd = false;

        // Iterate through the character counts
        for (int count : charCount.values()) {
            // If the count is even, we can include all characters
            if (count % 2 == 0) {
                longest_length += count;
            } else {
                // If it's odd, add its count - 1 to the longest length because an odd character
                // cannot form a palindrome
                longest_length += count - 1;

                // Indicates that have found an odd character
                odd = true;
            }
        }
        //If there's at least one odd character, we can place that character in the middle of the palindrome
        // So we add +1 to the length
        if (odd) {
            longest_length++;
        }
        return longest_length;
    }
    public static void main(String[] args) {
        HW8 hw8 = new HW8();
        String s = "abccccdd";
        s = s.toLowerCase();
        String s2 = "speediskey";
        s2 = s2.toLowerCase();
        System.out.println(hw8.longestPalindrome(s));
        System.out.println(hw8.longestPalindrome(s2));

        // Extra test case
        String s3 = "apple";
        s3 = s3.toLowerCase();
        System.out.println(hw8.longestPalindrome(s3));
    }
}